dask/array/creation.py

```diff
@@ -179,10 +199,14 @@ def full_like(a, fill_value, dtype=None, chunks=None):
 
     a = asarray(a, name=False)
     return full(
-        a.shape,
+        (shape or a.shape),
         fill_value,
         dtype=(dtype or a.dtype),
-        chunks=(chunks if chunks is not None else a.chunks),
+        chunks=(
+            "auto"
+            if shape and (chunks is None)
+            else (chunks if chunks is not None else a.chunks)
+        ),
```

Comment:
This is getting complex enough that I think I personally would understand it better if it was spelled out as a normal if-else block:

I think that this is what you're going for here, but I'm not sure.

```python
if shape is None:
    shape = a.shape
if chunks is None:
    if shape is None:
        chunks = a.chunks
    else:
        chunks = "auto"
```

Reply:
I concur